### DISCUSSION TITLE
zstd: make use of RLE_Block

### DISCUSSION THREAD

**Discussion by dsnet** [klauspost/compress#937] [Ideas/Features]
Consider the [following snippet](https://go.dev/play/p/xZXZZ9pzQ-s):
```go
bb := new(bytes.Buffer)
zw := must.Get(zstd.NewWriter(bb, zstd.WithEncoderLevel(zstd.SpeedBestCompression)))
must.Get(zw.Write(make([]byte, 10000)))
must.Do(zw.Close())
fmt.Printf("%x\n", bb.Bytes())
// 28b52ffd 64 1026 4d0000 080001540102310c87 0144e310
```
where we're compressing a input of 1e4 '\x00' bytes.

In the hexdump, we see that a Compressed_Block (type 2) was produced:
```
4d0000 080001540102310c87
```

Ideally, the compressor identify this as a candidate for RLE_Block (type 1):
```
833801 00
```
which is 8B shorter and faster to parse.

**Comment by klauspost** [comment]
Correct. There is no RLE detection. It is just such a fringe case it hasn't really come up.

It might be pretty cheap to detect after the compression loop. Let me see if can make it work reliably.

**Comment by klauspost** [comment]
Added in https://github.com/klauspost/compress/pull/938

**Reply by klauspost** [reply]
@dsnet Output of your example is:

```
// 28b52ffd 64 1026 4d0000 080001540102310c87 0144e310
-> 
// 28b52ffd 64 1026 83380100 0144e310
```

**Reply by dsnet** [reply]
Very cool. Thank you. Sparse files were one of the reasonable use cases that I thought this would be useful for.
